### PR TITLE
IncrementalIndex Tests and Benchmarks Parametrization

### LINKED ISSUE #10494: Remove redundant `IncrementalIndex.Builder`
### Motivation

#10335 added a per incremental-index builder, but the parent class builder (`IncrementalIndex.Builder`) was not removed to avoid 100+ lines changes in the test code.

### Proposed changes

Remove `IncrementalIndex.Builder` and refactor all its usage (only in tests the test code) to be parametrized (test both builder implementations (on-heap and off-heap).

### Rationale

Remove redundant code.

### Operational impact

This has no implication on the user experience.

### PR DESCRIPTION
**Pull request by liran-funaro**
Fixes #10494.

### Description

**Note:** This PR only affects tests and benchmarks.
It would help developers evaluate incremental-index extensions, such as oak-incremental-index (#10001).

#10335 added a per incremental-index builder, but the parent class builder (`IncrementalIndex.Builder`) was not removed to avoid 100+ line changes in the test code.
This PR removes `IncrementalIndex.Builder` and refactor all its usage (only in tests the test/benchmarks code).
In addition, where needed, a parametrization was added so it will test/benchmark both builder implementations (on-heap and off-heap).

#### Add test cases for each index type
All tests that are relevant to the incremental index were modified. The modifications include the parametrization of the tests for all incremental-index implementations: on-heap and off-heap. In addition, this PR includes a bug fix in `OffheapIncrementalIndex` that was found using these tests.

To support this, a new helper class was added: `IncrementalIndexCreator`.
This class handle creating the appropriate index according to its name and closing it at the end of each test.

#### Add benchmark cases for each index type
All the benchmarks that are relevant to the incremental index were added an incremental-index parametrization: on-heap or off-heap.
In addition, some of these benchmarks were modified to resolve some issues that were encountered.

We list here the additional modifications we made to some of the benchmarks.

  * Add some additional parametrization:
     - rollup opportunity for the row generator
     - number of rows per segment
     - query order: descending/ascending
  * Add a missing `tearDown()` procedure
  * Properly close the queryable index in the `tearDown()` procedure
  * Moved any temporary folder creation and deletion to the `setup()`/`tearDown()` methods so they would not affect the measurements of the results
  * Use a predefined seed for reproducible results, to be compliant with most benchmarks
  * Add scopes (`@State(Scope.Benchmark)`) that allow us to test the incremental index without the overhead of the setup procedure of the queryable index benchmark
     - One scope for benchmarking queries on the incremental index
     - One scope for benchmarking queries on the queryable index

In addition, to reduce code duplications, a few methods were added to `DataGenerator`:

1. `void addToIndex(IncrementalIndex<?> index, int numOfRows)`: adds rows from this generator to an existing index
1. `List<InputRow> toList(int numOfRows)`: adds rows from this generator to a new list

## User Experience Changes
After this PR, the user should not expect changes in most benchmarks results.
However, some benchmarks behavior will change as follows:

### Runtime
Expected change: the following benchmarks will run much faster due to eliminating redundancy setup/teardown procedure calls.
However, the benchmarks reported results should not change.
`FilteredAggregatorBenchmark`, `GroupByBenchmark`, `ScanBenchmark`, `SearchBenchmark`, `TimeseriesBenchmark`, `TopNBenchmark`

### Parametrization
Expected change: the following benchmarks will have additional parametrization options, hence they might take longer to run and produce more results.
* `indexType` parametrization (will also test the off-heap implementation): `FilteredAggregatorBenchmark`, `IncrementalIndexRowTypeBenchmark`, `IncrementalIndexReadBenchmark`, `IndexIngestionBenchmark`, `IndexPersistBenchmark`, `GroupByBenchmark`, `ScanBenchmark`, `SearchBenchmark`, `TimeseriesBenchmark`, `TopNBenchmark`
* `descending` query parametrization: `FilteredAggregatorBenchmark`, `TimeseriesBenchmark`
* `rollupOpportunity` ingestion parametrization: `IndexIngestionBenchmark`

### Unified Benchmarks Behaviour
Expected change: these changes affect some of the benchmarks' reported results as follows:

* A `rowsPerSegment` parametrization was added to `IncrementalIndexRowTypeBenchmark`. Before the number of rows was not parametrized and it reported the time per single row insertion. Now it reports the total insertion time of all the rows, like the rest of the tests report.
* `ScanBenchmark`, `SearchBenchmark`, `GroupByBenchmark`: now using a fixed seed, so the results are reproducible but might be slightly different than what was before with the random seed.
* `IndexPersistBenchmark`: this benchmark previously cleaned the temporary data folder inside the tested method, instead of in the teardown procedure. For large index sizes, it affected the benchmark result significantly. With this modification, the results will be different (shorter times), but it will better reflect the "persist" performance.

<hr>

This PR has:
- [X] been self-reviewed.
- [ ] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java [hunk 1]
```diff
@@ -205,30 +184,106 @@ public void setup() throws IOException
...
+  {
+    @Param({"onheap", "offheap"})
+    private String indexType;
+
+    IncrementalIndex<?> incIndex;
+
+    @Setup
+    public void setup(FilteredAggregatorBenchmark global) throws JsonProcessingException
+    {
+      global.appendableIndexSpec = IncrementalIndexCreator.parseIndexType(indexType);
```
Comment (Eshcar) [inline benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java:209]: worth mentioning that this is where the type of the index is set in the spec, and later used by the factory

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java [hunk 2]
```diff
@@ -149,39 +138,44 @@ public void teardown()
...
+  {
+    tmpDir = FileUtils.createTempDir();
+    log.info("Using temp dir: " + tmpDir.getAbsolutePath());
+  }
+
+  @TearDown(Level.Invocation)
+  public void teardownTemp() throws IOException
   {
-    return new IncrementalIndex.Builder()
+    FileUtils.deleteDirectory(tmpDir);
```
Comment (Eshcar) [inline benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java:151]: nice handling of the file

File: benchmarks/src/test/java/org/apache/druid/benchmark/query/GroupByBenchmark.java [hunk 3]
```diff
@@ -428,14 +424,12 @@ private void setupQueries()
...
 
   @Setup(Level.Trial)
-  public void setup() throws IOException
+  public void setup()
   {
     log.info("SETUP CALLED AT " + +System.currentTimeMillis());
 
     ComplexMetrics.registerSerde("hyperUnique", new HyperUniquesSerde());
 
-    executorService = Execs.multiThreaded(numProcessingThreads, "GroupByThreadPool[%d]");
```
Comment (Eshcar) [inline benchmarks/src/test/java/org/apache/druid/benchmark/query/GroupByBenchmark.java:437]: from this point onward a bit hard to follow the reasoning for the changes? what part of the PR description does this relate to?

File: processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java [hunk 4]
```diff
@@ -150,18 +150,13 @@ protected AddToFactsResult addToFacts(
...
-
     synchronized (this) {
       final AggregatorFactory[] metrics = getMetrics();
       final int priorIndex = facts.getPriorIndex(key);
       if (IncrementalIndexRow.EMPTY_ROW_INDEX != priorIndex) {
         final int[] indexAndOffset = indexAndOffsets.get(priorIndex);
-        bufferIndex = indexAndOffset[0];
-        bufferOffset = indexAndOffset[1];
-        aggBuffer = aggBuffers.get(bufferIndex).get();
+        ByteBuffer aggBuffer = aggBuffers.get(indexAndOffset[0]).get();
```
Comment (Eshcar) [inline processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:158]: what's the reasoning for these changes? add documentation to explain

File: processing/src/test/java/org/apache/druid/query/DoubleStorageTest.java [hunk 5]
```diff
@@ -321,10 +322,10 @@ private static QueryableIndex buildIndex(String storeDoubleAsFloat) throws IOExc
         )
         .build();
 
-    final IncrementalIndex index = new IncrementalIndex.Builder()
```
Comment (Eshcar) [inline processing/src/test/java/org/apache/druid/query/DoubleStorageTest.java:324]: from here on forward same 5 lines changes repeat for different tests

File: processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java [hunk 6]
```diff
@@ -268,7 +217,7 @@ private static MapBasedInputRow getLongRow(long timestamp, int dimensionCount)
   public void testCaseSensitivity() throws Exception
   {
     long timestamp = System.currentTimeMillis();
-    IncrementalIndex index = closerRule.closeLater(indexCreator.createIndex(DEFAULT_AGGREGATOR_FACTORIES));
+    IncrementalIndex<?> index = indexCreator.createIndex((Object) DEFAULT_AGGREGATOR_FACTORIES);
```
Comment (Eshcar) [inline processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java:220]: all changes from here are due to the generic type?

File: processing/src/test/java/org/apache/druid/segment/incremental/OffheapIncrementalIndexTestSpec.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.druid.collections.CloseableStupidPool;
+import org.apache.druid.utils.JvmUtils;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * Since the off-heap incremental index is not yet supported in production ingestion, we define its spec here only
```
Comment (Eshcar) [inline processing/src/test/java/org/apache/druid/segment/incremental/OffheapIncrementalIndexTestSpec.java:34]: add  a more general documentation of the role of this class for the time it is supported

**Review by Eshcar — COMMENTED** [review]
Thank Liran for this PR.

It is generally a very big PR.
The text description helps a lot in reviewing the code. 
However, there are some classes with very big changes, and it is not always easy to track back the reason for these changes, see my comments below.
Adding (many) lines of documentation within the code can greatly help the next reviewers read and approve the code.

File: benchmarks/src/test/java/org/apache/druid/benchmark/IncrementalIndexRowTypeBenchmark.java [hunk 8]
```diff
@@ -124,46 +131,51 @@ private MapBasedInputRow getStringRow(long timestamp, int dimensionCount)
...
       stringRows.add(getStringRow(0, DIMENSION_COUNT));
     }
   }
 
-  @Setup(Level.Iteration)
+  @Setup(Level.Invocation)
   public void setup2()
   {
     incIndex = makeIncIndex();
-    incFloatIndex = makeIncIndex();
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/IncrementalIndexRowTypeBenchmark.java:156]: Notice that the setup level was changed to per invocation, so for each benchmark, a new index is created.
There wasn't really a need for three different indices in the first place.

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilterPartitionBenchmark.java [hunk 9]
```diff
@@ -227,10 +228,10 @@ public void tearDown() throws IOException
 
   private IncrementalIndex makeIncIndex()
   {
-    return new IncrementalIndex.Builder()
+    return new OnheapIncrementalIndex.Builder()
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilterPartitionBenchmark.java:231]: I agree. But for the sake of reducing the diff size, I'd prefer to avoid this refactor.

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java [hunk 10]
```diff
@@ -205,30 +184,106 @@ public void setup() throws IOException
...
                   .build();
   }
 
-  @TearDown
-  public void tearDown() throws IOException
+  @State(Scope.Benchmark)
+  public static class IncrementalIndexState
+  {
+    @Param({"onheap", "offheap"})
+    private String indexType;
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java:202]: The idea here is indeed to allow a future index to be tested with the same code.
Using `Enum` will force this enumeration to list all existing index types in the core Druid package, albeit the index may only exist as an extension.
This way (using string), the user can choose any `indexType` name in the command line without it having to be pre-defined in the code.

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java [hunk 11]
```diff
@@ -205,30 +184,106 @@ public void setup() throws IOException
...
+    public void tearDown()
+    {
+      incIndex.close();
+    }
+  }
+
+  @State(Scope.Benchmark)
+  public static class IncrementalIndexIngestState
   {
-    FileUtils.deleteDirectory(tmpDir);
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java:223]: Note that it was not deleted. It just moved below to the teardown of the `QueryableIndexState`: `qIndexesDir.delete();`

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java (hunk 1 above)
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java:217]: Added.

File: benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java [hunk 12]
```diff
@@ -205,30 +184,106 @@ public void setup() throws IOException
...
+      );
+      incIndex.close();
+
+      qIndex = INDEX_IO.loadIndex(indexFile);
+    }
+
+    @TearDown
+    public void tearDown()
+    {
+      qIndex.close();
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/FilteredAggregatorBenchmark.java]: I'm not sure. But if setup fails, this might be an issue. Added a null check just to be safe.

File: benchmarks/src/test/java/org/apache/druid/benchmark/query/GroupByBenchmark.java (hunk 3 above)
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/query/GroupByBenchmark.java:437]: These changes are due to the scoping of the benchmark. This setup method now only in charge of initializing everything common for benchmarking both the incremental-index and the queriable-index.
Anything specific to the incremental or queriable index was moved to its designated scope below.

File: processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java (hunk 4 above)
Comment (liran-funaro) [reply processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:158]: Before this change, the code that responsible for the aggregation ran after a new row was inserted to `indexAndOffsets` (see line 209 below). This means that the new row was visible before any data was aggregated to it.
This does not correspond with the on-heap index behavior, which first aggregates the data, then inserts the row to the index.
According to `IncrementalIndexIngestionTest.testMultithreadAddFacts()`, the on-heap behavior is the correct one, so I changed it accordingly so the test will pass for this index as well.

File: processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java (hunk 6 above)
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java:220]: Yes. To allow testing any new incremental index.

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java (hunk 2 above)
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java:154]: Thanks!

File: processing/src/test/java/org/apache/druid/query/DoubleStorageTest.java (hunk 5 above)
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/query/DoubleStorageTest.java:324]: This is true for tests that do not test the incremental-index implementation. For such tests, we use the on-heap implementation because it is the most stable.
Tests that test the index itself, are now parametrized so they have more modifications other than these 5 lines.

File: processing/src/test/java/org/apache/druid/segment/incremental/OffheapIncrementalIndexTestSpec.java (hunk 7 above)
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/incremental/OffheapIncrementalIndexTestSpec.java]: Added.

**Comment by Eshcar** [comment]
Thanks for addressing the questions and issues 
LGTM

File: processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java (hunk 6 above)
Comment (Eshcar) [reply processing/src/test/java/org/apache/druid/segment/data/IncrementalIndexTest.java:220]: Thanks for addressing the questions and issues
LGTM

**Review by Eshcar — APPROVED** [review]
The changes in this PR would help people evaluate the oak extension.

**Review by a2l007 — COMMENTED** [review]
Thanks for the PR. I've left some comments. If you have the results from your local run of the modified benchmarks, could you please post them here?

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java [hunk 13]
```diff
@@ -110,28 +119,28 @@ public void setup() throws IOException
     );
 
     incIndex = makeIncIndex();
+    gen.addToIndex(incIndex, rowsPerSegment);
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java:122]: It wasn't planned, but I don't mind creating a follow-up PR for that or replacing everything in this PR.
What do you think is better?

File: processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java (hunk 4 above)
Comment (liran-funaro) [reply processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:158]: If this bug fix will be in a separate PR, then this PR will have a failing test.

File: processing/src/main/java/org/apache/druid/segment/generator/DataGenerator.java [hunk 14]
```diff
@@ -143,4 +151,54 @@ private long nextTimestamp()
...
+   * @param numOfRows the number of rows to generate
+   * @return a generator
+   */
+  private Stream<InputRow> generator(int numOfRows)
+  {
+    return Stream.generate(this::nextRow).limit(numOfRows);
+  }
+
+  /**
+   * Add rows form any generator to an index.
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/main/java/org/apache/druid/segment/generator/DataGenerator.java]: Fixed

File: processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+   * Generate an AppendableIndexSpec from index type.
+   *
+   * @param indexType an index type
+   * @return AppendableIndexSpec instance of this type
+   * @throws JsonProcessingException if failed to to parse the index
+   */
+  public static AppendableIndexSpec parseIndexType(String indexType) throws JsonProcessingException
+  {
+    return JSON_MAPPER.readValue(
+        String.format(Locale.ENGLISH, "{\"type\": \"%s\"}", indexType),
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java]: Fixed.

File: processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  /**
+   * Used to parameterize the tests with all the permutations of the parameters
+   * together with all the incremental index implementations.
+   *
+   * @param c a list of collections of parameters
+   * @return the cartesian product of all parameters
+   */
+  public static List<Object[]> indexTypeCartesianProduct(Collection<?>... c)
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java:200]: Added

File: processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+    return cartesianProduct(args);
+  }
+
+  /**
+   * Generates all the permutations of the parameters.
+   *
+   * @param c a list of collections of parameters
+   * @return the cartesian product of all parameters
+   */
+  public static List<Object[]> cartesianProduct(Collection<?>... c)
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java]: No. Changed to `private`.

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java [hunk 18]
```diff
@@ -110,28 +119,28 @@ public void setup() throws IOException
...
+  }
 
-    for (int j = 0; j < rowsPerSegment; j++) {
-      InputRow row = gen.nextRow();
-      if (j % 10000 == 0) {
-        log.info(j + " rows generated.");
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java:117]: I modified the `DataGenerator` to produce log events every 10,000 rows.

File: processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+import java.util.Locale;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * An incremental-index creator for parameterized incremental-index tests.
+ * It lists all the available incremental-index implementations, and responsible to create and close incremental-index
+ * instances during the tests.
+ */
+public class IncrementalIndexCreator implements Closeable
```
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/test/java/org/apache/druid/segment/incremental/IncrementalIndexCreator.java:56]: I added more detailed documentation for this class.

**Comment by liran-funaro** [comment]
@a2l007 Thanks for your review. I made modifications accordingly.
Let me know if you think more changes should be made, or if we can proceed to merge it.

I have results for the benchmarks I changed, but they are not with the default parameters.
I'll run the benchmarks overnight and post the results here.

**Comment by liran-funaro** [comment]
@a2l007 The benchmark run results are available here: https://pastebin.pl/view/raw/8a5559c7

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java [hunk 20]
```diff
@@ -87,101 +90,95 @@
   @Param({"true", "false"})
   private boolean rollup;
 
-  @Param({"none", "moderate", "high"})
-  private String rollupOpportunity;
+  @Param({"0", "1000", "10000"})
+  private int rollupOpportunity;
```
Comment (a2l007) [inline benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java:94]: Sorry I missed this earlier, but I feel that we should retain the textual values for rollupOpportunity as that is more user-friendly when reading the benchmark results.

**Review by a2l007 — COMMENTED** [review]
LGTM apart from one minor comment.
Also I'm curious if you've noticed any improvement in the benchmark scores even though the benchmark related changes are mainly refactoring and parametrization?

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java (hunk 20 above)
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IndexPersistBenchmark.java]: OK. I rolled back this change.

File: benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (liran-funaro) [reply benchmarks/src/test/java/org/apache/druid/benchmark/indexing/IncrementalIndexReadBenchmark.java:122]: OK. I'll publish a new PR for this once this PR is merged.

**Comment by liran-funaro** [comment]
Thanks, @a2l007. Please let me know what other modifications are required to approve this PR.

Regarding your question:

> Also I'm curious if you've noticed any improvement in the benchmark scores even though the benchmark-related changes are mainly refactoring and parameterization?

First, it is much faster to run the benchmarks due to less redundant setup/teardown procedure calls (achieved via scopes).
However, for most of the benchmarks, I didn't notice any performance changes.

The exceptions are as follows:

* `IndexPersistBenchmark`: this benchmark previously cleaned the temporary data folder inside the tested method, instead of in the teardown procedure. For large index sizes, it affected the benchmark result significantly.
With this modification, the results will be different (shorter times), but it will better reflect the "persist" performance.

* `IncrementalIndexRowTypeBenchmark`: before the number of rows was not parametrized and it reported the time per single row insertion. Now it reports the total insertion time of all the rows, like the rest of the tests report. There is no option in JMH to set the `@OperationsPerInvocation(MAX_ROWS)` with respect to the parameter `rowsPerSegment`.

* `ScanBenchmark`, `SearchBenchmark`: now using a fixed seed, so the results are reproducible but might be different than what was before with the random seed.

**Comment by a2l007** [comment]
> 

> * `IndexPersistBenchmark`: this benchmark previously cleaned the temporary data folder inside the tested method, instead of in the teardown procedure. For large index sizes, it affected the benchmark result significantly.
>   With this modification, the results will be different (shorter times), but it will better reflect the "persist" performance.
> 
> * `IncrementalIndexRowTypeBenchmark`: before the number of rows was not parametrized and it reported the time per single row insertion. Now it reports the total insertion time of all the rows, like the rest of the tests report. There is no option in JMH to set the `@OperationsPerInvocation(MAX_ROWS)` with respect to the parameter `rowsPerSegment`.
> 
> * `ScanBenchmark`, `SearchBenchmark`: now using a fixed seed, so the results are reproducible but might be different than what was before with the random seed.

I feel that this should be called out in the release notes so that users running benchmarks are aware of this change before upgrading. Could you please add a short description in the PR description mentioning the changes a user should expect running benchmarks before upgrading to this version. This would help the release manager add the blurb to the release notes.

**Review by a2l007 — APPROVED** [review]
Thanks for the PR!

**Comment by liran-funaro** [comment]
@a2l007 I added a "User Experience Changes" section in the PR description.
Thanks!

**Comment by liran-funaro** [comment]
Can we proceed to merge this PR?

**Comment by gianm** [comment]
> Can we proceed to merge this PR?

I haven't studied the patch, but the description makes sense to me (thanks for the detailed description), @a2l007 has reviewed it, and it seems pretty low risk (test/benchmark only) so I would say yes. Thanks for the contribution!

**Comment by liran-funaro** [comment]
Thanks!
